crates/xtask/src/changelog.rs

```diff
@@ -0,0 +1,377 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use anyhow::{ensure, Context, Result};
+use clap::ValueEnum;
+
+#[derive(clap::ValueEnum, Clone, Debug)]
+pub enum ChangelogReleaseType {
+    Major,
+    Minor,
+    Patch,
+}
+
+/// Changelog file data structure.
+#[derive(Debug, Default, PartialEq)]
+struct Changelog {
+    releases: Vec<ChangelogRelease>,
+
+    increment_count: u32,
+}
+
+impl Changelog {
+    fn push_release(&mut self, release: ChangelogRelease) {
+        self.releases.push(release);
+    }
+
+    fn set_increment_count(&mut self, num: u32) {
+        self.increment_count = num;
+    }
+}
+
+/// Each release of a Changelog, containing a version string and optional major, minor, patch change
+/// descriptions.
+#[derive(Debug, PartialEq)]
+struct ChangelogRelease {
+    version: String,
+
+    major: Option<Vec<String>>,
+    minor: Option<Vec<String>>,
+    patch: Option<Vec<String>>,
+}
+
+impl ChangelogRelease {
+    fn from(version: String) -> Self {
+        ChangelogRelease { version, major: None, minor: None, patch: None }
+    }
+
+    fn set_major(&mut self, contents: Vec<String>) {
+        self.major = Some(contents);
+    }
+
+    fn set_minor(&mut self, contents: Vec<String>) {
+        self.minor = Some(contents);
+    }
+
+    fn set_patch(&mut self, contents: Vec<String>) {
+        self.patch = Some(contents);
+    }
+}
+
+pub fn execute(crate_path: String, _version: ChangelogReleaseType, _message: String) -> Result<()> {
+    ensure!(
+        std::fs::exists(&crate_path).expect("Cannot open path right now."),
+        "Crate does not exist: {}",
+        crate_path
+    );
+
+    let _changelog = read_changelog_file(crate_path)?;
+    // TODO: act upon changelog...
+
+    Ok(())
+}
+
+fn read_changelog_file(crate_path: String) -> Result<Changelog> {
+    let changelog_path = format!("{crate_path}/CHANGELOG.md");
+
+    let changelog_contents = std::fs::read_to_string(&changelog_path)
+        .with_context(|| format!("reading {changelog_path}"))?;
+
+    parse_changelog(changelog_contents)
+}
+
+/// Converts raw file contents into a Changelog data structure.
+fn parse_changelog(contents: String) -> Result<Changelog> {
+    let mut changelog = Changelog::default();
+
+    // Trim off the trailing comment: <!-- Increment to skip ... -->
+    let contents: Vec<&str> =
+        contents.split("\n\n<!-- Increment to skip CHANGELOG.md test:").collect();
```

Comment:
We should probably use a state machine rather than string manipulations. Or at least we could use `contents.lines().peekable()` which should have enough state that we don't need anything more to know where we are.

For example we know the last version is `0.1.0` and has no sections and is directly followed by the skip counter.